yapapi/executor/__init__.py

```diff
@@ -557,22 +566,66 @@ async def start_worker(agreement: rest.market.Agreement, node_info: NodeInfo) ->
-                                    raise CommandExecutionError(evt.command, evt.message)
+                            async def get_batch_results():
+
+                                print(colors.yellow("[get_batch_results] Waiting for results..."))
 
-                            emit(events.GettingResults(agr_id=agreement.id, task_id=task_id))
-                            await batch.post()
-                            emit(events.ScriptFinished(agr_id=agreement.id, task_id=task_id))
-                            await accept_payment_for_agreement(agreement.id, partial=True)
+                                results = []
+                                async for evt_ctx in remote:
+                                    evt = evt_ctx.event(
+                                        agr_id=agreement.id, task_id=task_id, cmds=cmds
+                                    )
+                                    emit(evt)
+                                    results.append(evt)
+                                    if isinstance(evt, events.CommandExecuted) and not evt.success:
+                                        raise CommandExecutionError(evt.command, evt.message)
+
+                                emit(events.GettingResults(agr_id=agreement.id, task_id=task_id))
+                                await batch.post()
+                                emit(events.ScriptFinished(agr_id=agreement.id, task_id=task_id))
+                                await accept_payment_for_agreement(agreement.id, partial=True)
+
+                                print(colors.yellow("[get_batch_results] Got results"))
+                                return results
+
+                            if batch.wait_for_results or batch.contains_init_step:
+                                # Block until the results are available
+                                print(
+                                    colors.yellow(
+                                        "[executor] Blocking until results are available..."
+                                    )
+                                )
+                                results = await get_batch_results()
+                                print(colors.yellow("[executor] Returning results"))
+
+                                async def make_awaitable():
+                                    return results
+
+                                batch = await command_generator.asend(make_awaitable())
+                            else:
+                                # Schedule the coroutine in a separate asyncio task
+                                print(colors.yellow("[executor] Scheduling getting the results..."))
+                                loop = asyncio.get_event_loop()
+                                results_task = loop.create_task(get_batch_results())
+                                print(colors.yellow("[executor] Returning results task"))
+                                batch = await command_generator.asend(results_task)
```

Comment:
I believe you don't have to `make_awaitable` ... since `get_batch_results()` itself is an awaitable, isn't it?

```suggestion
                                # Block until the results are available
                                print(
                                    colors.yellow(
                                        "[executor] Blocking until results are available..."
                                    )
                                )
                                print(colors.yellow("[executor] Returning results"))
                                results = get_batch_results()
                            else:
                                # Schedule the coroutine in a separate asyncio task
                                print(colors.yellow("[executor] Scheduling getting the results..."))
                                loop = asyncio.get_event_loop()
                                results = loop.create_task(get_batch_results())
                                print(colors.yellow("[executor] Returning results task"))
                            batch = await command_generator.asend(results)
```

though, I still don't get how the first one is blocking since it returns an awaitable anyway ... will the `.commit()` indeed block in this case? ... I generally still don't grok this async magic fully :D

Reply:
ah, now I get it ... but I still think it won't be compatible with the current code -> iow, I guess there will be an exception saying that an awaitable was never awaited in case of the blocking call.... (won't there?)

Reply:
yeah, that's why I think trying to cover it with one method is not the best idea and it would be better to have _two_ commits:

* one (the current `.commit()`) would be always blocking and would _not_ return any results
* the other (`commit_async(wait_for_results: bool)`) would be either blocking or non-blocking but would always return an awaitable

Reply:
But it's not `commit()` that returns an awaitable, it's `yield`! `ctx.commit()` always returns a batch of commands (an instance `Work`). Anyway, returning awaitable in the first case, that is, when we're not interested in the results, is not a problem, in https://github.com/golemfactory/yapapi/pull/366/commits/6986d7bc0f2ccc964789b5f71417107b70c66b0a I'm wrapping the results in a future instead of a coroutine thus avoiding a warning about a coroutine that's never awaited.

Reply:
Therefore I think that just a single `commit()` method, with `blocking` defaulting to `True` should be enough.